File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js
// dev/stdin
const fs = require("fs");
let input = fs.readFileSync('dev/stdin').toString().split('\n')

let result = '';

for(let i = 0; i < input.length; i++) {
  let stack = [];
  const word = input[i];
  let isBalance = true;

  if(word == '.') break;

  for(let i = 0; i < word.length; i++) {
    if(word[i] == '(') {
      stack.push('(')
    } else if(word[i] == ')') {
      if(stack[stack.length-1] == '(') {
        stack.pop()
      } else {
        isBalance = false;
        break;
      }
    } else if(word[i] == '[') {
      stack.push('[')
    } else if(word[i] == ']') {
      if(stack[stack.length-1] == '[') {
        stack.pop()
      } else {
        isBalance = false;
        break;
      }
    }
  }
  if(stack.length > 0 || !isBalance) {
    result += 'no \n'
  } else {
    result += 'yes \n'
  }
}
console.log(result)

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
"""

"""
import sys


answer = ''

while True:
  word = sys.stdin.readline().rstrip()
  if word == '.' and len(word) == 1:
    break
  word = word.strip()

  stack = []
  result = 'yes'
  for s in word:

    if s == '(':
      stack.append('(')
    elif s == ')':
      if stack and stack[-1] == '(':
        stack.pop()
      elif len(stack) == 0 or stack[-1] != '(':
        result = 'no'
        break

    elif s == '[':
      stack.append('[')
    elif s == ']':
      if stack and stack[-1] == '[':
        stack.pop()
      elif len(stack) == 0 or stack[-1] != '[':
        result = 'no'
        break
  
  if len(stack) == 0 and result == 'yes':
    print('yes')
  else:
    print('no')
  
  
